### PR TITLE
Rework pyFAI calibrant as resource

### PR DESCRIPTION
**Pull request by vallsv**
This PR is a proposal to store the calibrant from pyFAI as a resource.

Plus some review of the code, documentation, typing.

Now a calibrant resource from pyfai is stored as `_filename = "pyfai:LaB6"`.
The file path is generated only when it have to be read.

This way it is easy to distinguish calibrant from resources, and other user calibrants.

If one try to change and save such calibrant, it raises an exception.

### CODE REVIEW

**Comment by vallsv** [comment]
This break the code if you think people use `.filename` to read files themselfs.

In this case i can propose another patch, to let behave like before.

File: pyFAI/calibrant.py [hunk 1]
```diff
@@ -580,22 +617,24 @@ def get_2th(self):
...
-        """Returns the index in the 2theta angle index
+    def get_2th_index(self, angle: float, delta: Optional[float]=None) -> int:
+        """Returns the index in the 2theta angle index.
 
         :param angle: expected angle in radians
         :param delta: precision on angle
         :return: 0-based index or None
         """
-        if angle and angle in self._2th:
+        if angle in self._2th:
```
Comment (kif) [inline pyFAI/calibrant.py:627]: issue are possible if angle is NaN

File: pyFAI/calibrant.py [hunk 2]
```diff
@@ -452,18 +475,19 @@ def _initialize(self):
...
+    def save_dSpacing(self, filename: Optional[str]=None):
         """
-        save the d-spacing to a file
-
+        Save the d-spacing to a file.
         """
         self._initialize()
         if (filename is None) and (self._filename is not None):
+            if self._filename.startswith("pyfai:"):
+                raise ValueError("A calibrant resource from pyFAI can(t be overwritten)")
```
Comment (kif) [inline pyFAI/calibrant.py:490]: Typo

File: pyFAI/calibrant.py (hunk 1 above)
Comment (vallsv) [reply pyFAI/calibrant.py:627]: I was curious about `bool(NaN)`
```
In [3]: if float("NaN"):
   ...:     print("Hi")
   ...: 
Hi
```

File: pyFAI/calibrant.py (hunk 2 above)
Comment (vallsv) [reply pyFAI/calibrant.py]: FIxed

**Comment by kif** [comment]
LGTM ... by the way, did you check the notebook on the documentation still run ?
or do they need dome adaptation ?

**Comment by vallsv** [comment]
I haven't checked anything.

The PR is not supposed to break anything, so tell me, if you are fine with the overall idea, i can rework the PR to make it work without regression.

**Pull request merged**